db/collection_index.go

```diff
@@ -0,0 +1,520 @@
+		}
+		indexes = append(indexes, collectionIndexDescription{
+			CollectionName: indexKey.CollectionName,
+			Index:          colDesk,
+		})
+	}
+
+	return indexes, nil
+}
+
+func (db *db) getCollectionIndexes(
+	ctx context.Context,
+	txn datastore.Txn,
+	colName string,
+) ([]client.IndexDescription, error) {
+	prefix := core.NewCollectionIndexKey(colName, "")
+	q, err := txn.Systemstore().Query(ctx, query.Query{
+		Prefix: prefix.ToString(),
+	})
+	if err != nil {
+		return nil, NewErrFailedToCreateCollectionQuery(err)
+	}
+	defer func() {
+		if err := q.Close(); err != nil {
+			log.ErrorE(ctx, "Failed to close collection query", err)
+		}
+	}()
+
+	indexes := make([]client.IndexDescription, 0)
+	for res := range q.Next() {
+		if res.Error != nil {
+			return nil, res.Error
+		}
+
+		var colDesk client.IndexDescription
+		err = json.Unmarshal(res.Value, &colDesk)
+		if err != nil {
+			return nil, NewErrInvalidStoredIndex(err)
+		}
+		indexes = append(indexes, colDesk)
+	}
+
+	return indexes, nil
+}
+
+func (c *collection) indexNewDoc(ctx context.Context, txn datastore.Txn, doc *client.Document) error {
+	indexes, err := c.getIndexes(ctx, txn)
+	if err != nil {
+		return err
+	}
+	for _, index := range indexes {
+		err = index.Save(ctx, txn, doc)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *collection) collectIndexedFields() []*client.FieldDescription {
+	fieldsMap := make(map[string]*client.FieldDescription)
+	for _, index := range c.indexes {
+		for _, field := range index.Description().Fields {
+			for i := range c.desc.Schema.Fields {
+				colField := &c.desc.Schema.Fields[i]
+				if field.Name == colField.Name {
+					fieldsMap[field.Name] = colField
+					break
+				}
+			}
+		}
+	}
+	fields := make([]*client.FieldDescription, 0, len(fieldsMap))
+	for _, field := range fieldsMap {
+		fields = append(fields, field)
+	}
+	return fields
+}
+
+func (c *collection) updateIndex(
+	ctx context.Context,
+	txn datastore.Txn,
+	doc *client.Document,
+) error {
+	_, err := c.getIndexes(ctx, txn)
+	if err != nil {
+		return err
+	}
+	oldDoc, err := c.get(ctx, txn, c.getPrimaryKeyFromDocKey(doc.Key()), c.collectIndexedFields(), false)
+	if err != nil {
+		return err
+	}
+	for _, index := range c.indexes {
+		err = index.Update(ctx, txn, oldDoc, doc)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *collection) CreateIndex(
+	ctx context.Context,
+	desc client.IndexDescription,
+) (client.IndexDescription, error) {
+	txn, err := c.getTxn(ctx, false)
+	if err != nil {
+		return client.IndexDescription{}, err
+	}
+
+	index, err := c.createIndex(ctx, txn, desc)
+	if err != nil {
+		return client.IndexDescription{}, err
+	}
+	if c.isIndexCached {
+		c.indexes = append(c.indexes, index)
+	}
+	err = c.indexExistingDocs(ctx, txn, index)
+	if err != nil {
+		return client.IndexDescription{}, err
+	}
+	return index.Description(), nil
+}
+
+func (c *collection) newFetcher() fetcher.Fetcher {
```

Comment:
question: I was just looking to pull this into my branch when I noticed - this feels like an odd location for such a function, particularly given that Lens was also looking to use it (created a lensed fetcher in `collection_index.go` feels extra odd).

At the moment it doesnt really do anything as far as I can tell so it is hard to really comment on it.  What is your plan for this function?

Reply:
It was just the simplest way i could impose a custom fetcher on a collection for testing purposes. It serves it's purpose but it's not very sophisticated. Didn't feel like doing more changes in this PR. I will probably create a task to address this.

Reply:
Okay no worries - can you please move it to another location though?  Maybe collection.go?